### PR TITLE
CSV Integration

### PR DESCRIPTION
**Pull request by nisarg73**

### CODE REVIEW

File: src/actions/genericActions.js [hunk 1]
```diff
@@ -146,3 +146,17 @@ export function handleUpdate(data, componentName) {
     rearrange: false
   };
 }
+
+export function handleCsvShow(componentName) {
+  return {
+    type: "HANDLE_CSV_SHOW" + "--" + componentName,
+    csv: true
```
Comment (algomaster99) [inline src/actions/genericActions.js:153]: Better to rename this to `csvVisibility`

File: src/components/csvIntegration.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
+import React, { Component } from "react";
+import { Segment, Button, Icon } from "semantic-ui-react";
+import { upperFirst } from "lodash"
+
+import { urlCsvIntegrtation } from '../urls';
+
+import style from "../styles.css";
+
+export class CsvIntegration extends Component {
```
Comment (algomaster99) [inline src/components/csvIntegration.js:9]: Rename it to https://github.com/IMGIITRoorkee/omniport-app-faculty-profile/pull/13#discussion_r454955633

File: src/components/csvIntegration.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          <Icon
+            color="grey"
+            name="delete"
+            onClick={() => this.props.handleCsvHide(componentName)}
+          />
+        </Segment>
+        <Segment style={{ width: "30vw" }} attached>
+        	<Button
+            color={appDetails.theme}
+            href={`${urlCsvIntegrtation()}?model=${this.state.modelName}`}
```
Comment (algomaster99) [inline src/components/csvIntegration.js:46]: @pradumangoyal told a better way for this. You can use `axios.get` to get the file content and download it using the JavaScript [File](https://developer.mozilla.org/en-US/docs/Web/API/File) class in one click.

File: src/constants/specs.js [hunk 4]
```diff
@@ -294,6 +298,7 @@ export const specs = {
   supervision: {
     icon: "star",
     draggable: true,
+    csv: true,
```
Comment (algomaster99) [inline src/constants/specs.js:301]: As of now, we need `csv: true` in only `Book` and `Paper` so make sure it is not committed.

File: src/urls.js [hunk 5]
```diff
@@ -5,3 +5,7 @@ export function urlMainApp() {
 export function urlCmsIntegration() {
   return `${urlMainApp()}cms/`;
 }
+
+export function urlCsvIntegrtation() {
```
Comment (algomaster99) [inline src/urls.js:9]: Change the name according to the class name suggested above.

File: src/components/genericListMaker.js [hunk 6]
```diff
@@ -95,6 +99,15 @@ const genericListMaker = (componentName, FormComponent) => {
                       }}
                     />
                   ) : null}
+                  {editMode &&
```
Comment (algomaster99) [inline src/components/genericListMaker.js:102]: What is `editMode` exactly? :sweat_smile:

File: src/components/csvIntegration.js (hunk 3 above)
Comment (pradumangoyal) [reply src/components/csvIntegration.js:46]: You may refer this https://github.com/IMGIITRoorkee/omniport-frontend-no-dues/blob/62bffbd1f613e2a88ab4d7ca71437587d0fa8639/src/components/Verifier/home.js#L294

File: src/components/genericListMaker.js (hunk 6 above)
Comment (nisarg73) [reply src/components/genericListMaker.js:102]: I am not sure why it was added by previous developers in the first place. Since its value has been hardcoded to `true` right from the start, and forever remains true only! :sweat:

**Comment by algomaster99** [comment]
Screenshot from 2020-07-17 02-51-15

Why are `_` appended and prepended to the name?

**Comment by nisarg73** [comment]
Sorry. This is due to the double quotes I used in filename attribute. I have removed those, you can pull the backend and this will be resolved.

File: src/components/csvIntegration.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        axios({
+            method: "post",
+            url: urlWriteAppendMultipleObjects(),
+            data: data,
+            headers: headers
+        }).then(response => {
+          this.props.handleUpdate(response.data, this.props.componentName);
+          this.props.handleCsvHide(this.props.componentName);
+        }).catch(error => {
+            if (error.response.status == "400") {
```
Comment (algomaster99) [inline src/components/csvIntegration.js:121]: Please google the difference between `==` and `===`. You will get why this line of code is so egregious.

File: src/components/csvIntegration.js [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+      }
+    });
+  };
+
+  handleErrors = error_dict => {
+    let dict = error_dict;
+    let errors = [];
+    for (let key in dict) {
+      for (let index in dict[key]) {
+        errors.push(startCase(key) + ": " + dict[key][index]);
```
Comment (algomaster99) [inline src/components/csvIntegration.js:97]: meme

**Comment by pradumangoyal** [comment]
- Excel icon should be before the add icon

**Comment by pradumangoyal** [comment]
Nice work @nisarg73 🎉

**Comment by nisarg73** [comment]
@pradumangoyal Regarding last point, When I tried to serve .xlsx and .xls file via download, and upload data using it, `read_csv()` was able to parse and process those files. Which is why I included them in the message.

But just now I tried uploading file made from excel in windows on staging, it doesn't work :disappointed:

**Comment by algomaster99** [comment]
@nisarg73 Make the changes above and complete the checklist. Then we are good to go.

File: src/actions/genericActions.js [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import axios from "axios";
 import { initial } from "../constants/initial";
 import { specs } from "../constants/specs";
+import { urlGetAffordances } from "../urls"
```
Comment (algomaster99) [inline src/actions/genericActions.js:4]: Add `;`

File: src/actions/genericActions.js [hunk 10]
```diff
@@ -146,3 +147,36 @@ export function handleUpdate(data, componentName) {
...
+export function fetchAffordances(componentName) {
+  return (dispatch) => {
+    axios.get(urlGetAffordances(), {
+      params: {
+        model: componentName
+      }
+    })
+    .then(response => {
+      dispatch({
+        type: "FETCH_AFFORDANCES" + "--" + componentName,
```
Comment (algomaster99) [inline src/actions/genericActions.js:174]: You can use `` `FETCH_AFFORDANCES--${componentName}` `` but don't change it here. Will break consistency but keep this in mind for later projects :smile:

File: src/components/csvIntegration.js [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+        "fileLink": null,
+        "uploadType": "append",
+      },
+      errors: [],
+      loading: false,
+    };
+  }
+
+  componentDidMount() {
+    document.addEventListener("keydown", this.handleKeyPress, false);
```
Comment (algomaster99) [inline src/components/csvIntegration.js:40]: Pressing <kbd>Enter</kbd> doesn't submit the sheet for me.

File: src/components/csvIntegration.js [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+      }
+    }
+    this.setState({ errors: errors });
+  };
+
+  handleSubmit = () => {
+    this.setState({ loading: true });
+    if (this.state.data.fileLink == null) {
+      this.setState({ 
+        errors: ["You need to upload file before submitting."],
```
Comment (algomaster99) [inline src/components/csvIntegration.js:119]: Can we somehow get this error from backend?

File: src/components/csvIntegration.js [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+            color="grey"
+            name="delete"
+            onClick={() => this.props.handleCsvHide(componentName)}
+          />
+        </Segment>
+        <Segment attached styleName="style.formStyle">
+          <ErrorTransition errors={errors} />
+          <Message
+            info
+            header="Column headers has to satisfy below shown order and properties."
```
Comment (algomaster99) [inline src/components/csvIntegration.js:175]: - ~has~ have
- the below

I think we should rephrase it as: "Column headers have to satisfy the below-stated constraints."

File: src/components/csvIntegration.js [hunk 14]
```diff
@@ -0,0 +1,249 @@
...
+            list={instructions}
+          />
+          <Form autoComplete="off">
+            {data.uploadType == "append"
+              ? (
+                <Message compact size="tiny">
```
Comment (algomaster99) [inline src/components/csvIntegration.js:181]:
```jsx
<Message compact size="tiny">
  Append: Data from file will be added without changing{' '}
  existing items.
</Message>
```

Reference: https://stackoverflow.com/a/41142436/11751642

File: src/components/csvIntegration.js [hunk 15]
```diff
@@ -0,0 +1,249 @@
...
+                  Append: Data from file will be added without changing existing items.
+                </Message>
+              )
+              : (
+                <Message compact size="tiny" color="yellow">
+                  <Icon name="warning sign" />
+                  New: All the existing items will be deleted first before adding new.
```
Comment (algomaster99) [inline src/components/csvIntegration.js:188]: Same as above

File: src/components/csvIntegration.js [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+              )
+            }
+            <Form.Group inline>
+              <label>Type: </label>
+              <Form.Field>
+                <Radio
+                  label="Append"
+                  name="uploadType"
+                  value="append"
+                  checked={data.uploadType == "append"}
```
Comment (algomaster99) [inline src/components/csvIntegration.js:199]: Instil in your mind to **always** use `===`.

File: src/constants/specs.js [hunk 17]
```diff
@@ -1287,19 +1301,6 @@ export const specs = {
         },
         user_props: ["handleChange"]
       },
-      {
```
Comment (algomaster99) [inline src/constants/specs.js:1290]: Why did you delete this?

File: src/styles.css [hunk 18]
```diff
@@ -25,6 +25,9 @@
 .formMinWidth {
   min-width: 475px;
 }
+.csvMinWidth {
+  min-width: 480px;
```
Comment (algomaster99) [inline src/styles.css:29]: Make this `475px` too?

File: src/constants/specs.js (hunk 17 above)
Comment (nisarg73) [reply src/constants/specs.js:1290]: Because `paper` model doesn't have a `contribution` field.
